### DISCUSSION TITLE
Explanation of tags & version numbers

### DISCUSSION THREAD

**Discussion by kristof-mattei** [devcontainers/community#109] [Q&A] [answered]
Hi all,

Trying to understand the components of the build numbers.

For example at https://github.com/devcontainers/images/tree/main/src/rust:

This mentions the the following tags:

```
mcr.microsoft.com/devcontainers/rust:latest (or bookworm, bullseye, buster to pin to an OS version)
mcr.microsoft.com/devcontainers/rust:1 (or 1-bookworm, 1-bullseye, 1-buster to pin to an OS version)
```

Questions: 
* Which version number the `-1` part referring to here? 
* If we omit the `bookworm` part, which version does it pull? (and where is this defined).

I searched the [the dockerfile guide](https://containers.dev/guide/dockerfile) but could not find the answers to these questions.

Moving further, we notice that this image is versioned as [1.0.7](https://github.com/devcontainers/images/blob/main/src/rust/history/1.0.7.md), which in turn is built from  [v0.3.27](https://github.com/devcontainers/images/tree/v0.3.27/src/rust). 

What is the relationship between all these versions? What is the reason to keep them separate? 

Thanks

**Comment by samruddhikhandale** [comment]
Hi 👋 

Every image is built for multiple different OS flavors (eg. bookworm, bullseye, focal, jammy etc). Hence, you can pin your image to your choice of OS. Then for semantic versioning, https://github.com/devcontainers/images/blob/main/src/rust/manifest.json#L2 points to the version of the image. 

> Which version number the -1 part referring to here?

We recommend users to pin to `major` versions in order to receive bug fixes, security patches, improvements etc.
In case of rust, `1-` refers to latest published image, it will continue to receive minor and patch versions automatically.

> If we omit the bookworm part, which version does it pull? (and where is this defined).

The default is `bookworm`, hence you will end up with a `bookworm` distro. See https://github.com/devcontainers/images/blob/main/src/rust/manifest.json#L9

> Moving further, we notice that this image is versioned as [1.0.7](https://github.com/devcontainers/images/blob/main/src/rust/history/1.0.7.md), which in turn is built from [v0.3.27](https://github.com/devcontainers/images/tree/v0.3.27/src/rust).

`1.0.7` is the version of the image and [v0.3.27](https://github.com/devcontainers/images/tree/v0.3.27/src/rust) is the version of the release. Feel free to ignore release versions because we use that for tracking image releases.

https://github.com/devcontainers/images/tree/main/src/rust/history has release docs published for every image version, this is helpful for comparing changes and keeping track of image contents.

Let us know if anything is unclear, happy to clarify any other doubts.

**Reply by kristof-mattei** [reply]
Thank you, that's clear!

Looking at the Image Version itself:

https://github.com/devcontainers/images/blob/main/src/rust/history/1.0.7.md

For the Variant: bookworm:

```
mcr.microsoft.com/devcontainers/rust:1.0.7-bookworm
mcr.microsoft.com/devcontainers/rust:1.0.7-1
mcr.microsoft.com/devcontainers/rust:1.0.7-1-bookworm
```

In the last line we have the `1.0.7` component which is the image version (which is what you referred to as the part which should use as `1` to get all updates.

What is then the part after the dash, the `-1`, `-1-bookworm` and `-bookworm`. I understand `-bookworm`, but what is the difference between `-1-bookworm` and `-bookworm`? 

(I'm assuming that `-1` is the same as `-1-bookworm` because of the `latest` bit you mentioned in the `json` config.)

**Reply by samruddhikhandale** [reply, marked as answer]
> mcr.microsoft.com/devcontainers/rust:1.0.7-1-bookworm

- `1.0.7` is the rust's image version
- `1` is the major version of `Rust` language. (eg. the current version of rust is `1.75.0` - hence, major version is `1`)

It might be simpler to visualize with Python, see https://github.com/devcontainers/images/blob/main/src/python/history/1.1.4.md#variant-312-bookworm

> mcr.microsoft.com/devcontainers/python:1.1.4-3.12-bookworm

- 1.1.4 : Image version
- 3.12 : Version of Python
- bookworm - OS distro

@kristof-mattei Does that make sense?

**Reply by kristof-mattei** [reply]
Yes! Much clearer now! Thanks!
